### PR TITLE
Optimised generation of fresh names.

### PR DESCRIPTION
**Pull request by fajb**
The current `Tactics.fresh_id_in_env` calls
`ids_of_named_context_val : ids_of_named_context_val : named_context_val -> Id.Set.t`
which is a linear operation.

Instead, we pass the predicate
`val ids_of_named_context_val_mem : named_context_val -> Id.t -> bool`
which checks whether an identifier is within the map.

The rest of the modifications are consequences of the API change.

In Termpops, we still explicitly constructs an intermediate Id.Set.t in
`vars_of_env(_mem)` for the `rel_context` but replace
`Context.Rel.fold_outside` by `Context.Rel.fold_inside`.

**Kind:** performance

### CODE REVIEW

**Review by Matafou — APPROVED** [review]
LGTM concerning funind.

**Comment by SkySkimmer** [comment]
Started bench https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/834/console

**Comment by fajb** [comment]
The failure in `equations` looks benign and it seems the fix would be backward compatible 
by using directly `Tactics.fresh_id_in_env` and `Tactics.fresh_id` instead of implementing them.

**Comment by SkySkimmer** [comment]
~~~
┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬────────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]      │    mem faults     │
│                             │                         │                                             │                                             │                                │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD  PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1353.54 1369.92 -1.20 % │     3772809532382     3819983145344 -1.23 % │     6765537845052     6765700792243 -0.00 % │    1293900    1292980  +0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-bignums │   62.45   63.08 -1.00 % │      173458570120      174889127279 -0.82 % │      243862229257      244060504482 -0.08 % │     483356     483236  +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   47.61   48.06 -0.94 % │      131120946601      131439077816 -0.24 % │      171144465576      171174879942 -0.02 % │     556244     556252  -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-performance-tests │   12.08   12.18 -0.82 % │       33282665806       33402260588 -0.36 % │       57066628593       57134088186 -0.12 % │     454624     462860  -1.78 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-hott │  388.42  391.62 -0.82 % │     1060004276530     1067335274237 -0.69 % │     1733357963053     1733485636917 -0.01 % │     630040     630112  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fiat-core │  101.11  101.89 -0.77 % │      287968664982      288379818747 -0.14 % │      388283398541      388343723131 -0.02 % │     496856     497752  -0.18 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │  976.19  983.35 -0.73 % │     2718851480353     2740254571122 -0.78 % │     4446456673132     4446201747277 +0.01 % │    1022272     930164  +9.90 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2094.90 2109.77 -0.70 % │     5846776294462     5888457002241 -0.71 % │    11486024798474    11486241217340 -0.00 % │     951664     951468  +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-rewriter │  568.25  571.89 -0.64 % │     1576958048122     1587379427251 -0.66 % │     2440391517395     2440447710360 -0.00 % │     996336    1000592  -0.43 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-fiat-crypto-legacy │ 6013.92 6052.32 -0.63 % │    16739525945026    16851295470724 -0.66 % │    28585824664677    28584522083711 +0.00 % │    2371312    2354988  +0.69 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-unimath │ 4017.05 4042.67 -0.63 % │    11188427270709    11262242376687 -0.66 % │    21441691744211    21439288683722 +0.01 % │    1109504     979840 +13.23 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-coqutil │   58.26   58.62 -0.61 % │      162507870416      163183949108 -0.41 % │      208901939997      209275733612 -0.18 % │     534504     534524  -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1524.72 1533.66 -0.58 % │     4252075982677     4270508935362 -0.43 % │     6166767998643     6168403842823 -0.03 % │    1142192    1142244  -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-math-classes │  204.87  205.80 -0.45 % │      572318830575      574329972394 -0.35 % │      768331788711      768313756911 +0.00 % │     517056     516696  +0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-verdi │  118.50  119.03 -0.45 % │      327058938649      328712948924 -0.50 % │      426613501252      426628777970 -0.00 % │     542448     543644  -0.22 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  256.52  257.57 -0.41 % │      711168534507      714606007749 -0.48 % │     1216842182759     1216712004265 +0.01 % │    1206924    1207092  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│             coq-fiat-crypto │ 4231.00 4246.27 -0.36 % │    11749712598520    11790764476669 -0.35 % │    19383737684502    19386742408881 -0.02 % │    2355004    2353772  +0.05 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-corn │ 1420.67 1425.29 -0.32 % │     3961558983894     3976664854348 -0.38 % │     6152113371006     6152134679464 -0.00 % │     871020     871088  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  175.87  176.41 -0.31 % │      489583051212      491165334779 -0.32 % │      696448904886      696523146262 -0.01 % │     792548     792500  +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  144.92  145.36 -0.30 % │      403601822014      404315112124 -0.18 % │      535727705778      535719950651 +0.00 % │     613280     613196  +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-flocq │  206.25  206.84 -0.29 % │      574189292329      574567274097 -0.07 % │      740791625944      740958599573 -0.02 % │     927204     847264  +9.44 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1519.14 1523.46 -0.28 % │     4238609158046     4254471977892 -0.37 % │     6512912276604     6512524183592 +0.01 % │    1380988    1384744  -0.27 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   49.06   49.19 -0.26 % │      136424854520      136537188554 -0.08 % │      189576936819      189586290027 -0.00 % │     557672     557768  -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-compcert │  720.68  721.96 -0.18 % │     2008393796562     2011165493384 -0.14 % │     2901245861403     2901299018964 -0.00 % │    1039796    1039740  +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  263.12  263.51 -0.15 % │      732134897797      733588918197 -0.20 % │     1123488967138     1123466627650 +0.00 % │     789312     790068  -0.10 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  164.18  164.19 -0.01 % │      455966453958      456453152907 -0.11 % │      637215252870      637203341438 +0.00 % │     848860     824164  +3.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-color │  648.16  647.98 +0.03 % │     1809613844484     1810872609799 -0.07 % │     2309751059073     2310170128176 -0.02 % │    1140324    1139072  +0.11 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  597.95  597.51 +0.07 % │     1672721929521     1672743058517 -0.00 % │     2692721707443     2693549511024 -0.03 % │    3170360    3168796  +0.05 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                  coq-sf-plf │   43.58   43.51 +0.16 % │      121712263246      121842440325 -0.11 % │      168642205431      168764509975 -0.07 % │     494672     494936  -0.05 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-coqprime │   84.52   84.37 +0.18 % │      231868825310      232124639567 -0.11 % │      334699112642      334660917891 +0.01 % │     756376     756080  +0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1272.18 1269.79 +0.19 % │     3544350302524     3542240492715 +0.06 % │     5025103596514     5025535829747 -0.01 % │    1819384    1819656  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-coquelicot │   73.52   73.37 +0.20 % │      201428562535      201981966786 -0.27 % │      266556586430      266564410014 -0.00 % │     655232     654852  +0.06 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴────────────────────────────────┴───────────────────┘
~~~

**Comment by SkySkimmer** [comment]
Needs an overlay for equations. Fiat crypto failure is spurious, as we will see when the overlay is pushed.

File: engine/termops.mli [hunk 1]
```diff
@@ -227,6 +227,7 @@ val env_rel_context_chop : int -> env -> env * rel_context
 
 (** Set of local names *)
 val vars_of_env: env -> Id.Set.t
+val vars_of_env_mem: env -> (Id.t -> bool)
```
Comment (ppedrot) [inline engine/termops.mli]: I don't think that this function belongs here. It should be moved to `Environ` instead.

File: kernel/environ.mli [hunk 2]
```diff
@@ -145,7 +145,7 @@ val named_context_of_val : named_context_val -> Constr.named_context
 val val_of_named_context : Constr.named_context -> named_context_val
 val empty_named_context_val : named_context_val
 val ids_of_named_context_val : named_context_val -> Id.Set.t
-
+val ids_of_named_context_val_mem : named_context_val -> Id.t -> bool
```
Comment (ppedrot) [inline kernel/environ.mli]: The function name doesn't follow the conventions from that file, and furthermore it doesn't match its actual behaviour. I'd rather call it `mem_var_val`, and similarly for the environ-taking variant, `mem_var`.

File: engine/termops.mli (hunk 1 above)
Comment (fajb) [reply engine/termops.mli]: Do you mean both?

Comment (ppedrot) [reply engine/termops.mli]: Technically both should be there, but for some historical reason the above one was put in that file. I don't know whether you should move the other one in the same PR, though.

File: kernel/environ.mli (hunk 2 above)
Comment (ppedrot) [reply kernel/environ.mli]: (Also, arguments are flipped w.r.t. the standard convention, see e.g. `mem_constant.)

Comment (fajb) [reply kernel/environ.mli]: I'll do it thanks.
To be sure, you mean `mem_var_val : named_context_val -> Id.t -> bool` or
`mem_var_named_context_val : named_context_val -> Id.t -> bool`

Comment (ppedrot) [reply kernel/environ.mli]: The first one.

**Comment by fajb** [comment]
commit https://github.com/coq/coq/pull/11397/commits/4f93146ec389918a1313853b57ea7e2f6d0bf215 implements the discussed moving/renaming of functions

commit https://github.com/coq/coq/pull/11397/commits/a6dd03efdadec8afca49a427beb007f9ecbd7b10 is an attempt at using an Id.AvoidSet.t abstraction.
The changes are not negligable. There are a few places where this avoids building an Id.Set.t from the context...

Nonetheless, I would be curious whether this has some non-negliable positive efficiency impact.

**Comment by ppedrot** [comment]
Rereading this PR, I am not sure that the added code complexity is worth it when considering the performance gains. I'd recommend closure of the PR in favour of other lia-specific enhancements.

**Comment by fajb** [comment]
I am OK with that.

**Pull request closed without merging**